libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -488,20 +488,28 @@ method publish*(g: GossipSub,
 
   var peers: HashSet[PubSubPeer]
 
+  # add always direct peers
+  peers.incl(g.explicit.getOrDefault(topic))
+
+  if topic in g.topics: # if we're subscribed use the mesh
+    peers.incl(g.mesh.getOrDefault(topic))
+
   if g.parameters.floodPublish:
+    let
+      msgSize = data.len
+      bandwidth = 25_000_000 #TODO replace with bandwidth estimate
+      msToTransmit = max(msgSize div (bandwidth div 1000), 1)
+      maxFloodPublish =
```

Comment:
no need for a `min`?

Reply:
consider the case where we're not subscribed to the topic - sending to 0 peers seems a bit harsh then...

Reply:
Right, that won't happen before >50mb, but still good to cover it

Reply:
thought a bit more about this, and I think we want to put the minimum at dmin at least .. 

it feels very risky to send only to one peer - if that peer is slow, it'll delay the message for all peers by a heartbeat _and_ it relies a bit too heavily on the IHAVE/IWANT mechanism to recover, ie there's no redundancy...

I know there's a cost here for sending big messages from slow peers, but I think the risk for normal/highbandwidth peers is more real and costly.

Reply:
this also implies that rather than treating this as a per-send thing, we should have a global rate limiter that triggers for all sends, specially iwants so that we don't succumb to iwant-overload while doing parallel sending here

Reply:
The minimum is currently dLow (it will be caught by the `if peers.len < g.parameters.dLow:` below)

Reply:
aren't fanout peers populated only when we're not subscribed? also as such, there might not be enough of them in the fanout table either?

Reply:
We also use them when we can't publish to enough peers (ie bad mesh), and they are replenished when < dLow

Reply:
> We also use them when we can't publish to enough peers (ie bad mesh), and they are replenished when < dLow


so sending to `g.gossipsub[topic]` as a last resort wouldn't be appropriate?

Reply:
This should be the same as using the fanout, the fanout is only a small cache of `gossipsub` to have more stable diffusion routes, basically